In [1]:
import pandas as pd
import numpy as np
from sklearn import linear_model

# Multiple linear regression

In [2]:
df = pd.read_excel('M_Regression.xlsx')
df

Area  Bedrooms  Age_house     Price
0   2000         3         20   5500000
1   3500         4         15   8650000
2   1200         2         18   4500000
3   3600         3         30   7950000
4   4000         5          8  12600000
5   2300         2         12   2300000
6   3100         3         18  14500000
7   2900         4         35   3000000
8   5000         6         26  13400000
9   3400         5         10   8009000
10  4500         8         13  17600000
11  1500         2          5   3800000
12  1700         3          1   5000000
13  1000         1         12   3500000
14  1400         2         30   6800000
15  2300         3         12   9000000
16  1500         2          1   4300000
17  2400         3         12   2300000
18  5000         6         16  23400000
19  3200         2          1  12600000
20  2300         2         11   2300000
21  3200         3         18  14500000
22  1200         2          2   4500000
23  3600         3         30   7950000
24  4000         5          4  22600000
25  1700         2          1   5000000
26  1000         4         10   2500000
27  1400         3         12   7800000
28  2300         5         13   9000000
29  1500         5          1   5300000
30  2000         1          2   5800000
31  1300         2         11   2300000
32  3200         3         10  13500000
33  1200         2         12   4500000
34  3600         3         12   7950000

In [3]:
import statsmodels.formula.api as smf
formula="Price~Area+Bedrooms+Age_house "
reg1=smf.ols(formula,df).fit()
print(reg1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Price   R-squared:                       0.652
Model:                            OLS   Adj. R-squared:                  0.619
Method:                 Least Squares   F-statistic:                     19.38
Date:                Tue, 26 Sep 2023   Prob (F-statistic):           2.89e-07
Time:                        17:53:30   Log-Likelihood:                -574.11
No. Observations:                  35   AIC:                             1156.
Df Residuals:                      31   BIC:                             1162.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -9.1e+05   1.55e+06     -0.585      0.5

From regression we can say : how much good is our test that can be checked by R2 value
here R2 value is close to one means our prediction is very good
some cases when we are doing regression our varibles we choosen p value > 0.05 that cases we reove that values wth p value >0.05 and again rerun the analysis...that time we again get a r2 value and everything...

Result:
R2 value of this house price prediction data is close to 1 so our prediction is avery good prediction
durbin whatson value of regressiob should be always close to 2 here durbin whatson value is so close to 2... we got a intercept value from that we can predict other houses prices also

by equation to predict area 3000, bedrooms 6, age 5

In [4]:
-9.1e+05+(3459.3236*3000)+(5.425e+05*6)+(-1.209e+05*5)             

#beta0 + Beta1 *x1 + beta2 *x2 + beta 3* x3 + epsilon

12118470.8

here we predicted house price with 3000 sq feet area bedrooms 6 and age of house 5 years old house so its price of the house is 12118470.8

In [5]:
#area=1700, bedroom =4, age =15

-9.1e+05+(3459.3236*1700)+(5.425e+05*4)+(-1.209e+05*15) 

5327350.12

# another way to find answer

In [6]:
dfq = pd.read_excel('M_Regression.xlsx')
dfq

Area  Bedrooms  Age_house     Price
0   2000         3         20   5500000
1   3500         4         15   8650000
2   1200         2         18   4500000
3   3600         3         30   7950000
4   4000         5          8  12600000
5   2300         2         12   2300000
6   3100         3         18  14500000
7   2900         4         35   3000000
8   5000         6         26  13400000
9   3400         5         10   8009000
10  4500         8         13  17600000
11  1500         2          5   3800000
12  1700         3          1   5000000
13  1000         1         12   3500000
14  1400         2         30   6800000
15  2300         3         12   9000000
16  1500         2          1   4300000
17  2400         3         12   2300000
18  5000         6         16  23400000
19  3200         2          1  12600000
20  2300         2         11   2300000
21  3200         3         18  14500000
22  1200         2          2   4500000
23  3600         3         30   7950000
24  4000         5          4  22600000
25  1700         2          1   5000000
26  1000         4         10   2500000
27  1400         3         12   7800000
28  2300         5         13   9000000
29  1500         5          1   5300000
30  2000         1          2   5800000
31  1300         2         11   2300000
32  3200         3         10  13500000
33  1200         2         12   4500000
34  3600         3         12   7950000

In [7]:
X = dfq[['Area', 'Bedrooms', 'Age_house']]
X


Area  Bedrooms  Age_house
0   2000         3         20
1   3500         4         15
2   1200         2         18
3   3600         3         30
4   4000         5          8
5   2300         2         12
6   3100         3         18
7   2900         4         35
8   5000         6         26
9   3400         5         10
10  4500         8         13
11  1500         2          5
12  1700         3          1
13  1000         1         12
14  1400         2         30
15  2300         3         12
16  1500         2          1
17  2400         3         12
18  5000         6         16
19  3200         2          1
20  2300         2         11
21  3200         3         18
22  1200         2          2
23  3600         3         30
24  4000         5          4
25  1700         2          1
26  1000         4         10
27  1400         3         12
28  2300         5         13
29  1500         5          1
30  2000         1          2
31  1300         2         11
32  3200         3         10
33  1200         2         12
34  3600         3         12

In [8]:
y = dfq['Price']
y

0      5500000
1      8650000
2      4500000
3      7950000
4     12600000
5      2300000
6     14500000
7      3000000
8     13400000
9      8009000
10    17600000
11     3800000
12     5000000
13     3500000
14     6800000
15     9000000
16     4300000
17     2300000
18    23400000
19    12600000
20     2300000
21    14500000
22     4500000
23     7950000
24    22600000
25     5000000
26     2500000
27     7800000
28     9000000
29     5300000
30     5800000
31     2300000
32    13500000
33     4500000
34     7950000
Name: Price, dtype: int64

In [9]:
reg = linear_model.LinearRegression()
reg.fit(X, y)


LinearRegression()

In [10]:
reg.coef_

array([   3459.32364699,  542475.02759242, -120864.61485264])

In [11]:
 reg.predict([[1700, 4, 15]])

C:\Users\kchri\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([5327807.69508921])

In [12]:
 reg.predict([[2700, 2, 5]])

C:\Users\kchri\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([8910827.43542115])

In [13]:
reg.predict([[3000, 6, 25]])

C:\Users\kchri\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([9701232.34283521])

### example 2

## Data preprocessing (null values by median value)

In [14]:
df1 = pd.read_excel('M_Regression_2.xlsx')
df1

Area  Bedrooms  Age_house     Price
0   2000       3.0         20   5500000
1   3500       4.0         15   8650000
2   1200       2.0         18   4500000
3   3600       NaN         30   7950000
4   4000       5.0          8  12600000
5   2300       2.0         12   2300000
6   3100       3.0         18  14500000
7   2900       4.0         35   3000000
8   5000       6.0         26  13400000
9   3400       5.0         10   8009000
10  4500       NaN         13  17600000
11  1500       2.0          5   3800000
12  1700       3.0          1   5000000
13  1000       1.0         12   3500000
14  1400       2.0         30   6800000
15  2300       3.0         12   9000000
16  1500       NaN          1   4300000
17  2400       3.0         12   2300000
18  5000       6.0         16  23400000
19  3200       2.0          1  12600000
20  2300       2.0         11   2300000
21  3200       NaN         18  14500000
22  1200       2.0          2   4500000
23  3600       3.0         30   7950000
24  4000       5.0          4  22600000
25  1700       2.0          1   5000000
26  1000       4.0         10   2500000
27  1400       3.0         12   7800000
28  2300       5.0         13   9000000
29  1500       NaN          1   5300000
30  2000       1.0          2   5800000
31  1300       2.0         11   2300000
32  3200       3.0         10  13500000
33  1200       2.0         12   4500000
34  3600       3.0         12   7950000

In [15]:
df1.Bedrooms.median()

3.0

In [16]:
df1.Bedrooms = df1.Bedrooms.fillna(df1.Bedrooms.median())
df1

Area  Bedrooms  Age_house     Price
0   2000       3.0         20   5500000
1   3500       4.0         15   8650000
2   1200       2.0         18   4500000
3   3600       3.0         30   7950000
4   4000       5.0          8  12600000
5   2300       2.0         12   2300000
6   3100       3.0         18  14500000
7   2900       4.0         35   3000000
8   5000       6.0         26  13400000
9   3400       5.0         10   8009000
10  4500       3.0         13  17600000
11  1500       2.0          5   3800000
12  1700       3.0          1   5000000
13  1000       1.0         12   3500000
14  1400       2.0         30   6800000
15  2300       3.0         12   9000000
16  1500       3.0          1   4300000
17  2400       3.0         12   2300000
18  5000       6.0         16  23400000
19  3200       2.0          1  12600000
20  2300       2.0         11   2300000
21  3200       3.0         18  14500000
22  1200       2.0          2   4500000
23  3600       3.0         30   7950000
24  4000       5.0          4  22600000
25  1700       2.0          1   5000000
26  1000       4.0         10   2500000
27  1400       3.0         12   7800000
28  2300       5.0         13   9000000
29  1500       3.0          1   5300000
30  2000       1.0          2   5800000
31  1300       2.0         11   2300000
32  3200       3.0         10  13500000
33  1200       2.0         12   4500000
34  3600       3.0         12   7950000

In [17]:
import statsmodels.formula.api as smf
formula="Q('Price')~Area+Bedrooms+Age_house "
reg1=smf.ols(formula,df1).fit()
print(reg1.summary())

                            OLS Regression Results                            
Dep. Variable:             Q('Price')   R-squared:                       0.646
Model:                            OLS   Adj. R-squared:                  0.612
Method:                 Least Squares   F-statistic:                     18.87
Date:                Tue, 26 Sep 2023   Prob (F-statistic):           3.76e-07
Time:                        17:53:31   Log-Likelihood:                -574.41
No. Observations:                  35   AIC:                             1157.
Df Residuals:                      31   BIC:                             1163.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -9.353e+05   1.67e+06     -0.561      0.5

In [18]:
df=pd.read_excel('')
df

FileNotFoundError: [Errno 2] No such file or directory: ''